File: app/rcc/scripts/ubx_ublox.js
/**
 * UBX Protocol (u-blox) Parser
 *
 * Parses UBX binary protocol messages from u-blox GNSS receivers.
 *
 * UBX is a proprietary binary protocol used by u-blox GPS/GNSS modules.
 * It provides comprehensive navigation data, satellite information, and
 * receiver configuration. More efficient than NMEA with higher update rates.
 *
 * INPUT FORMAT: Binary UBX message frame
 * OUTPUT ARRAY: Extracted navigation data based on message class and ID
 *
 * Note: This parser requires Binary (Direct) decoder mode in the Project Editor.
 *       Frame delimiters (0xB5 0x62 sync characters) are automatically removed
 *       by Serial Studio.
 */

//------------------------------------------------------------------------------
// Configuration
//------------------------------------------------------------------------------

/**
 * Total number of values in the output array.
 * This should match the number of datasets you've configured in Serial Studio.
 */
const numItems = 20;

/**
 * Maps UBX message classes and IDs to parsing configurations.
 *
 * UBX messages are identified by Class + ID combination.
 *
 * Common UBX message types:
 *   0x01 0x07: NAV-PVT (Position Velocity Time Solution)
 *   0x01 0x35: NAV-SAT (Satellite Information)
 *   0x01 0x06: NAV-SOL (Navigation Solution)
 *   0x01 0x02: NAV-POSLLH (Position LLH)
 *   0x01 0x12: NAV-VELNED (Velocity NED)
 *   0x01 0x21: NAV-TIMEUTC (UTC Time Solution)
 *
 * Message 0x01 0x07 (NAV-PVT) is most comprehensive and commonly used.
 */
const messageMap = {
  "0x01-0x07": {
    // NAV-PVT: position, velocity, time in a single message
    indices: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    parser: function(data) {
      var iTOW = extractUint32LE(data, 0);

      var longitude = extractInt32LE(data, 24) * 1e-7;
      var latitude  = extractInt32LE(data, 28) * 1e-7;
      var height    = extractInt32LE(data, 32) * 0.001;
      var hMSL      = extractInt32LE(data, 36) * 0.001;

      var velN    = extractInt32LE(data, 48) * 0.001;
      var velE    = extractInt32LE(data, 52) * 0.001;
      var velD    = extractInt32LE(data, 56) * 0.001;
      var gSpeed  = extractInt32LE(data, 60) * 0.001;
      var headMot = extractInt32LE(data, 64) * 1e-5;

      var numSV = data[23];

      return [latitude, longitude, hMSL, height, gSpeed, headMot,
              velN, velE, velD, numSV, iTOW];
    }
  },
  "0x01-0x35": {
    // NAV-SAT: satellite information summary
    indices: [11, 12, 13],
    parser: function(data) {
      var iTOW    = extractUint32LE(data, 0);
      var version = data[4];
      var numSvs  = data[5];
      return [iTOW, version, numSvs];
    }
  },
  "0x01-0x06": {
    // NAV-SOL: ECEF position, accuracy estimate, satellite count
    indices: [14, 15, 16, 17, 18],
    parser: function(data) {
      var iTOW  = extractUint32LE(data, 0);
      var ecefX = extractInt32LE(data, 12) * 0.01;
      var ecefY = extractInt32LE(data, 16) * 0.01;
      var ecefZ = extractInt32LE(data, 20) * 0.01;
      var pAcc  = extractUint32LE(data, 24) * 0.01;
      var numSV = data[47];
      return [ecefX, ecefY, ecefZ, pAcc, numSV];
    }
  },
  "0x01-0x02": {
    // NAV-POSLLH: geodetic position solution
    indices: [0, 1, 2, 3],
    parser: function(data) {
      var lon    = extractInt32LE(data, 4) * 1e-7;
      var lat    = extractInt32LE(data, 8) * 1e-7;
      var height = extractInt32LE(data, 12) * 0.001;
      var hMSL   = extractInt32LE(data, 16) * 0.001;
      return [lat, lon, hMSL, height];
    }
  }
};

/**
 * Array to hold all parsed values.
 * Values persist between frames unless updated by new data.
 */
const parsedValues = new Array(numItems).fill(0);

//------------------------------------------------------------------------------
// Helper Functions
//------------------------------------------------------------------------------

/**
 * Extracts a 16-bit unsigned integer from byte array (little-endian).
 * UBX protocol uses little-endian byte order.
 */
function extractUint16LE(bytes, offset) {
  return bytes[offset] | (bytes[offset + 1] << 8);
}

/**
 * Extracts a 32-bit unsigned integer from byte array (little-endian).
 */
function extractUint32LE(bytes, offset) {
  return bytes[offset] | (bytes[offset + 1] << 8) |
         (bytes[offset + 2] << 16) | (bytes[offset + 3] << 24);
}

/**
 * Extracts a 32-bit signed integer from byte array (little-endian).
 */
function extractInt32LE(bytes, offset) {
  var value = bytes[offset] | (bytes[offset + 1] << 8) |
              (bytes[offset + 2] << 16) | (bytes[offset + 3] << 24);
  return value > 2147483647 ? value - 4294967296 : value;
}

/**
 * Validates the UBX Fletcher's checksum.
 * Covers class, ID, length, and payload bytes.
 */
function validateChecksum(msgClass, msgId, length, payload, ckA, ckB) {
  var calcCkA = 0, calcCkB = 0;

  calcCkA = (calcCkA + msgClass) & 0xFF; calcCkB = (calcCkB + calcCkA) & 0xFF;
  calcCkA = (calcCkA + msgId) & 0xFF;    calcCkB = (calcCkB + calcCkA) & 0xFF;
  calcCkA = (calcCkA + (length & 0xFF)) & 0xFF;        calcCkB = (calcCkB + calcCkA) & 0xFF;
  calcCkA = (calcCkA + ((length >> 8) & 0xFF)) & 0xFF; calcCkB = (calcCkB + calcCkA) & 0xFF;

  for (var i = 0; i < payload.length; i++) {
    calcCkA = (calcCkA + payload[i]) & 0xFF;
    calcCkB = (calcCkB + calcCkA) & 0xFF;
  }

  return (calcCkA === ckA) && (calcCkB === ckB);
}

//------------------------------------------------------------------------------
// Frame Parser Function
//------------------------------------------------------------------------------

/**
 * Parses a UBX message frame into the predefined array structure.
 *
 * UBX message structure:
 *   Bytes 0-1:  Sync characters (0xB5 0x62)
 *   Byte 2:     Message class
 *   Byte 3:     Message ID
 *   Bytes 4-5:  Payload length (little-endian)
 *   Bytes 6-n:  Payload
 *   Byte n+1:   Checksum A (CK_A)
 *   Byte n+2:   Checksum B (CK_B)
 *
 * Note: Serial Studio removes sync characters, so this parser receives:
 * [class][id][len_lo][len_hi][payload...][ck_a][ck_b]
 *
 * How it works:
 * 1. Extracts message class and ID
 * 2. Extracts payload length
 * 3. Validates checksum
 * 4. Looks up parser for this class+ID combination
 * 5. Calls appropriate parser
 * 6. Stores values at their mapped indices
 *
 * @param {array} frame - The binary UBX frame from the data source (byte array)
 * @returns {array} Array of values mapped according to messageMap
 */
function parse(frame) {
  if (frame.length < 6)
    return parsedValues;

  var msgClass     = frame[0];
  var msgId        = frame[1];
  var payloadLength = frame[2] | (frame[3] << 8);

  if (frame.length < 4 + payloadLength + 2) {
    console.error("Incomplete UBX frame");
    return parsedValues;
  }

  var payload = [];
  for (var i = 0; i < payloadLength; i++)
    payload.push(frame[4 + i]);

  var ckA = frame[4 + payloadLength];
  var ckB = frame[4 + payloadLength + 1];

  if (!validateChecksum(msgClass, msgId, payloadLength, payload, ckA, ckB)) {
    console.error("UBX checksum validation failed");
    return parsedValues;
  }

  var messageKey = "0x" + msgClass.toString(16).padStart(2, '0') +
                   "-0x" + msgId.toString(16).padStart(2, '0');

  if (!messageMap.hasOwnProperty(messageKey))
    return parsedValues;

  var config = messageMap[messageKey];

  try {
    var values = config.parser(payload);
    for (var i = 0; i < values.length && i < config.indices.length; i++)
      parsedValues[config.indices[i]] = values[i];
  }
  catch (e) {
    console.error("UBX parsing error:", e.message);
  }

  return parsedValues;
}

File: app/rcc/scripts/parser/lua/ubx_ublox.lua
--
-- UBX Protocol (u-blox) Parser
--
-- Parses UBX binary protocol messages from u-blox GNSS receivers.
--
-- UBX is a proprietary binary protocol used by u-blox GPS/GNSS modules.
-- It provides comprehensive navigation data, satellite information, and
-- receiver configuration. More efficient than NMEA with higher update rates.
--
-- INPUT FORMAT: Binary UBX message frame
-- OUTPUT ARRAY: Extracted navigation data based on message class and ID
--
-- Note: This parser requires Binary (Direct) decoder mode in the Project Editor.
--       Frame delimiters (0xB5 0x62 sync characters) are automatically removed
--       by Serial Studio. The parser receives:
--         [class][id][len_lo][len_hi][payload...][ck_a][ck_b]
--

--------------------------------------------------------------------------------
-- Configuration
--------------------------------------------------------------------------------

-- Total number of values in the output array.
-- This should match the number of datasets you've configured in Serial Studio.
local numItems = 20

-- Array to hold all parsed values.
-- Values persist between frames unless updated by new data.
local parsedValues = {}
for i = 1, numItems do parsedValues[i] = 0 end

--------------------------------------------------------------------------------
-- Helper Functions (zero-based offsets, matching JS reference)
--------------------------------------------------------------------------------

-- Extracts a 16-bit unsigned integer from a byte table (little-endian).
-- UBX protocol uses little-endian byte order.
local function extractUint16LE(bytes, offset)
  local b0 = bytes[offset + 1] or 0
  local b1 = bytes[offset + 2] or 0
  return b0 | (b1 << 8)
end

-- Extracts a 32-bit unsigned integer from a byte table (little-endian).
local function extractUint32LE(bytes, offset)
  local b0 = bytes[offset + 1] or 0
  local b1 = bytes[offset + 2] or 0
  local b2 = bytes[offset + 3] or 0
  local b3 = bytes[offset + 4] or 0
  return b0 | (b1 << 8) | (b2 << 16) | (b3 << 24)
end

-- Extracts a 32-bit signed integer from a byte table (little-endian).
local function extractInt32LE(bytes, offset)
  local v = extractUint32LE(bytes, offset)
  if v > 2147483647 then v = v - 4294967296 end
  return v
end

-- Validates the UBX Fletcher's checksum.
-- Covers class, ID, length, and payload bytes.
local function validateChecksum(msgClass, msgId, length, payload, ckA, ckB)
  local calcA, calcB = 0, 0

  calcA = (calcA + msgClass) & 0xFF;           calcB = (calcB + calcA) & 0xFF
  calcA = (calcA + msgId)    & 0xFF;           calcB = (calcB + calcA) & 0xFF
  calcA = (calcA + (length & 0xFF)) & 0xFF;    calcB = (calcB + calcA) & 0xFF
  calcA = (calcA + ((length >> 8) & 0xFF)) & 0xFF; calcB = (calcB + calcA) & 0xFF

  for i = 1, #payload do
    calcA = (calcA + payload[i]) & 0xFF
    calcB = (calcB + calcA) & 0xFF
  end

  return (calcA == ckA) and (calcB == ckB)
end

--------------------------------------------------------------------------------
-- Message Parsers
--------------------------------------------------------------------------------

-- NAV-PVT (0x01 0x07): Position, Velocity, Time solution.
-- Returns 11 values matching JS: lat, lon, hMSL, height, gSpeed, headMot,
-- velN, velE, velD, numSV, iTOW.
local function parseNavPvt(payload)
  local iTOW      = extractUint32LE(payload, 0)
  local longitude = extractInt32LE(payload, 24) * 1e-7
  local latitude  = extractInt32LE(payload, 28) * 1e-7
  local height    = extractInt32LE(payload, 32) * 0.001
  local hMSL      = extractInt32LE(payload, 36) * 0.001
  local velN      = extractInt32LE(payload, 48) * 0.001
  local velE      = extractInt32LE(payload, 52) * 0.001
  local velD      = extractInt32LE(payload, 56) * 0.001
  local gSpeed    = extractInt32LE(payload, 60) * 0.001
  local headMot   = extractInt32LE(payload, 64) * 1e-5
  local numSV     = payload[24] or 0

  return {latitude, longitude, hMSL, height, gSpeed, headMot,
          velN, velE, velD, numSV, iTOW}
end

-- NAV-SAT (0x01 0x35): Satellite information summary.
-- Returns iTOW, version, numSvs.
local function parseNavSat(payload)
  local iTOW    = extractUint32LE(payload, 0)
  local version = payload[5] or 0
  local numSvs  = payload[6] or 0
  return {iTOW, version, numSvs}
end

-- NAV-SOL (0x01 0x06): ECEF position, accuracy estimate, satellite count.
local function parseNavSol(payload)
  local iTOW  = extractUint32LE(payload, 0)
  local ecefX = extractInt32LE(payload, 12) * 0.01
  local ecefY = extractInt32LE(payload, 16) * 0.01
  local ecefZ = extractInt32LE(payload, 20) * 0.01
  local pAcc  = extractUint32LE(payload, 24) * 0.01
  local numSV = payload[48] or 0
  return {ecefX, ecefY, ecefZ, pAcc, numSV}
end

-- NAV-POSLLH (0x01 0x02): Geodetic position solution.
local function parseNavPosllh(payload)
  local lon    = extractInt32LE(payload, 4)  * 1e-7
  local lat    = extractInt32LE(payload, 8)  * 1e-7
  local height = extractInt32LE(payload, 12) * 0.001
  local hMSL   = extractInt32LE(payload, 16) * 0.001
  return {lat, lon, hMSL, height}
end

-- Maps UBX message class+ID combinations to parser function and output indices.
local messageMap = {
  ["0x01-0x07"] = { indices = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}, parser = parseNavPvt    },
  ["0x01-0x35"] = { indices = {12, 13, 14},                         parser = parseNavSat    },
  ["0x01-0x06"] = { indices = {15, 16, 17, 18, 19},                 parser = parseNavSol    },
  ["0x01-0x02"] = { indices = {1, 2, 3, 4},                         parser = parseNavPosllh }
}

--------------------------------------------------------------------------------
-- Frame Parser Function
--------------------------------------------------------------------------------

-- Parses a UBX message frame into the predefined array structure.
-- @param frame The binary UBX frame (byte table, sync chars removed)
-- @return Array of values mapped according to messageMap
function parse(frame)
  -- Need at least class + id + length(2) + 2 checksum bytes
  if #frame < 6 then
    return parsedValues
  end

  -- Extract header fields
  local msgClass      = frame[1]
  local msgId         = frame[2]
  local payloadLength = frame[3] | (frame[4] << 8)

  -- Validate frame is long enough for declared payload + checksum
  if #frame < 4 + payloadLength + 2 then
    return parsedValues
  end

  -- Copy payload into a zero-based-addressable table (indices 1..N)
  local payload = {}
  for i = 1, payloadLength do
    payload[i] = frame[4 + i]
  end

  -- Extract checksum bytes
  local ckA = frame[4 + payloadLength + 1]
  local ckB = frame[4 + payloadLength + 2]

  -- Validate UBX Fletcher checksum
  if not validateChecksum(msgClass, msgId, payloadLength, payload, ckA, ckB) then
    return parsedValues
  end

  -- Build the class+ID lookup key
  local messageKey = string.format("0x%02x-0x%02x", msgClass, msgId)
  local config = messageMap[messageKey]
  if not config then
    return parsedValues
  end

  -- Invoke the matching parser and store values at their mapped indices
  local values = config.parser(payload)
  for i = 1, #values do
    local idx = config.indices[i]
    if idx then
      parsedValues[idx] = values[i]
    end
  end

  return parsedValues
end
